### PR TITLE
ddl: add truncate partition all support

### LINKED ISSUE #22704: TRUNCATE PARTITION ALL gets stuck, ddl_worker retries, error returned
### 1. Minimal reproduce step (Required)

```sql
create table t (a int primary key) partition by hash (a);
alter table t truncate partition all;
```

### 2. What did you expect to see? (Required)

Speed up execution and return _understandable_ error messages. Now it runs for several minutes (5min+) and fails.

### 3. What did you see instead (Required)

````
[2021/02/03 10:27:26.695 +00:00] [INFO] [error.go:25] ["ERROR 1735: DDL JOB ROLLBACK, ERROR MSG: UNKNOWN PARTITION 'DROP?' IN TABLE 'TABLE_5
_BINARY_2'"]
````

You should be able to see log like the following:

````
[2021/02/03 11:32:00.966 +00:00] [ERROR] [ddl_worker.go:720] ["[ddl] run DDL job error"] [worker="worker 3, tp general"] [error="[table:1735]Unknown partition 'drop?' in table 'table_5_binary_4'"]
[2021/02/03 11:32:00.975 +00:00] [INFO] [ddl_worker.go:499] ["[ddl] run DDL job failed, sleeps a while then retries it."] [worker="worker 3, tp general"] [waitTime=1s] [error="[table:1735]Unknown partition 'drop?' in table 'table_5_binary_4'"]
[2021/02/03 11:32:01.975 +00:00] [INFO] [ddl_worker.go:774] ["[ddl] schema version doesn't change"] [worker="worker 3, tp general"]
[2021/02/03 11:32:01.976 +00:00] [INFO] [ddl_worker.go:599] ["[ddl] run DDL job"] [worker="worker 3, tp general"] [job="ID:1198, Type:truncate partition, State:running, SchemaState:none, SchemaID:1190, TableID:1192, RowCount:0, ArgLen:0, start time: 2021-02-03 11:31:09.291 +0000 UTC, Err:[table:1735]Unknown partition 'drop?' in table 'table_5_binary_4', ErrCount:52, SnapshotVersion:0"]
[2021/02/03 11:32:01.978 +00:00] [ERROR] [ddl_worker.go:720] ["[ddl] run DDL job error"] [worker="worker 3, tp general"] [error="[table:1735]Unknown partition 'drop?' in table 'table_5_binary_4'"]
[2021/02/03 11:32:01.986 +00:00] [INFO] [ddl_worker.go:499] ["[ddl] run DDL job failed, sleeps a while then retries it."] [worker="worker 3, tp general"] [waitTime=1s] [error="[table:1735]Unknown partition 'drop?' in table 'table_5_binary_4'"]
````

### 4. What is your TiDB version? (Required)

````
-----------------------------------+
| Release Version: v5.0.0-rc
Edition: Community
Git Commit Hash: fd4437dde445012f5b75e5d4f336433cb076c1ee
Git Branch: heads/refs/tags/v5.0.0-rc
UTC Build Time: 2021-01-12 06:00:03
GoVersion: go1.13
Race Enabled: false
TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306
Check Table Before Drop: false |
+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
1 row in set (0.00 sec)
````

### PR DESCRIPTION
**Pull request by AndrewDi**
### What problem does this PR solve?

Issue Number: close #22704

Problem Summary:

### What is changed and how it works?

What's Changed:

if ddl contains `truncate partition all`, all current partition id will pass to `ddl_api`

How it Works:

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test

Side effects

### Release note 
- alter table truncate partition now support `all` option

### CODE REVIEW

**Comment by AndrewDi** [comment]
releated parser pr is [1179](https://github.com/pingcap/parser/pull/1179)

**Comment by xhebox** [comment]
@AndrewDi You need a rebase. There are too many commits coauthored by you.

**Comment by AndrewDi** [comment]
@xhebox rebase has been done, sorry for the later response.

File: ddl/db_test.go [hunk 1]
```diff
@@ -6606,6 +6606,7 @@ func (s *testSerialDBSuite) TestIssue22819(c *C) {
 	tk1.MustExec("update t1 set v = 2 where v = 1")
 
 	tk2.MustExec("alter table t1 truncate partition p0")
+	tk2.MustExec("alter table t1 truncate partition all")
```
Comment (xhebox) [inline ddl/db_test.go:6609]: I think you should add a standalone test, which does not interfere with this test.

**Review by xhebox — COMMENTED** [review]
Rest LGTM, but it is a new feature. You should not close the issue until the document is updated.

File: ddl/db_test.go (hunk 1 above)
Comment (bb7133) [reply ddl/db_test.go:6609]: Agreed, please add a test case named like 'TestTruncateAllPartitions' and check the the result of this syntax.

Thanks!

Comment (bb7133) [reply ddl/db_test.go:6609]: Agreed, please add a test case named like 'TestTruncateAllPartitions' and check the the result of this syntax.

Thanks!

**Comment by AndrewDi** [comment]
@xhebox @bb7133 PTAL, add test case

File: ddl/db_test.go (hunk 1 above)
Comment (AndrewDi) [reply ddl/db_test.go]: done, please take a look

File: ddl/db_test.go [hunk 2]
```diff
@@ -6606,8 +6606,20 @@ func (s *testSerialDBSuite) TestIssue22819(c *C) {
...
+
+func (s *testSerialSuite) TestTruncateAllPartitions(c *C) {
+	tk1 := testkit.NewTestKit(c, s.store)
+	tk1.MustExec("use test;")
+	tk1.MustExec("drop table if exists partition_table;")
+	defer func() {
+		tk1.MustExec("drop table if exists partition_table;")
+	}()
+	tk1.MustExec("create table partition_table (v int) partition by hash (v) partitions 2;")
+	tk1.MustExec("insert into partition_table values (1);")
```
Comment (xhebox) [inline ddl/db_test.go:6622]: We can make a better test that emphasizes on **all** partitions:

1. partitioning by less than 10 and 20
2. insert 8 and 15, to insert into all partitions
3. truncate all partitions and `count(*) == 0`

Comment (AndrewDi) [reply ddl/db_test.go]: PTAL again?

Comment (xhebox) [reply ddl/db_test.go]: It is fine to me. Plz fix the build, and do not close the issue. I think it is better to leave the issue open for documentation PR.

**Comment by xhebox** [comment]
There is a failed test in check_dev_2, could you investigate whether it is because of your modifications? @AndrewDi

**Comment by AndrewDi** [comment]
> There is a failed test in check_dev_2, could you investigate whether it is because of your modifications? @AndrewDi

No, I didn't touch privileges part, but I will fix the test case, may be some one has modified parser parted, but tidb part has not merged.

**Comment by xhebox** [comment]
> No, I didn't touch privileges part, but I will fix the test case, may be some one has modified parser parted, but tidb part has not merged.

OK, here is the [PR](https://github.com/pingcap/parser/commit/0118b271c7760207026fffc7dd918787d5fe8ca7). And your fix is correct. But I am not sure if this should be fixed in this PR... I will ask in our slack channel.

File: executor/revoke_test.go [hunk 3]
```diff
@@ -82,7 +82,7 @@ func (s *testSuite1) TestRevokeTableScope(c *C) {
 
 	// Make sure all the table privs for new user is Y.
 	res := tk.MustQuery(`SELECT Table_priv FROM mysql.tables_priv WHERE User="testTblRevoke" and host="localhost" and db="test" and Table_name="test1"`)
-	res.Check(testkit.Rows("Select,Insert,Update,Delete,Create,Drop,Index,Alter"))
+	res.Check(testkit.Rows("Select,Insert,Update,Delete,Create,Drop,Index,Alter,Show View"))
```
Comment (xiongjiwei) [inline executor/revoke_test.go:85]: it this true in MySQL? I haven't tested it.

Comment (xhebox) [reply executor/revoke_test.go:85]: From mysql-8.0.22

```
| Select,Insert,Update,Delete,Create,Drop,References,Index,Alter,Create View,Show view,Trigger |
``

Comment (xiongjiwei) [reply executor/revoke_test.go:85]: we are not supporting `trigger` now, I think it ok here.

**Review by xiongjiwei — APPROVED** [review]

**Comment by xhebox** [comment]
@AndrewDi Please fix the conflicts. And I will fire the merge.

**Comment by AndrewDi** [comment]
> @AndrewDi Please fix the conflicts. And I will fire the merge.

done...please merge asap

**Pull request merged**